apm-sniffer/apm-sdk-plugin/finagle-6.25.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/finagle/AnnotationInterceptor.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.skywalking.apm.plugin.finagle;
+
+import org.apache.skywalking.apm.agent.core.context.ContextManager;
+import org.apache.skywalking.apm.agent.core.context.trace.AbstractSpan;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.EnhancedInstance;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.MethodInterceptResult;
+
+import java.lang.reflect.Method;
+
+import static org.apache.skywalking.apm.plugin.finagle.ContextHolderFactory.getLocalContextHolder;
+import static org.apache.skywalking.apm.plugin.finagle.FinagleCtxs.getContextCarrier;
+import static org.apache.skywalking.apm.plugin.finagle.FinagleCtxs.getSpan;
+
+/**
+ * Finagle use Annotation to represent data that tracing system interested, usually these annotations are created by
+ * filters after ClientTracingFilter in the rpc call stack. We can intercept annotations that we interested.
+ */
+public class AnnotationInterceptor {
+
+    abstract static class Abstract extends AbstractInterceptor {
+
+        @Override
+        public void onConstructImpl(EnhancedInstance enhancedInstance, Object[] objects) {
+            onConstruct(enhancedInstance, objects, getSpan());
+        }
+
+        protected abstract void onConstruct(EnhancedInstance enhancedInstance, Object[] objects, AbstractSpan span);
+    }
+
+    /**
+     * When we create exitspan in ClientTracingFilter, we can't know the operation name, however the Rpc annotation
+     * contains the operation name we need, so we intercept the constructor of this Annotation and set operation name
+     * to exitspan.
+     */
+    public static class Rpc extends Abstract {
+
+        @Override
+        protected void onConstruct(EnhancedInstance enhancedInstance, Object[] objects, AbstractSpan span) {
+            if (objects != null && objects.length == 1) {
+                String rpc = (String) objects[0];
+                if (span == null) {
+                    // in case the exitspan is created later
```

Comment:
Iniitally, here is used to set op to server entryspan(sorry for the wrong comments),  and finally, I passed op from client to server, so it is no longer needed here. The codes and comments has been modified as below:
```java
if (span != null) {
    /*
     * The Rpc Annotation is created both in client side and server side, in server side, this
     * annotation is created only in finagle versions below 17.12.0.
     *
     * If the span is not null, which means we are in the client side, we just set op to the exitspan.
     *
     * If the span is null, which means we are in the server side with finagle version below 17.12.0.
     * In server side, we don't need this annotation, because we can get op from Contexts.broadcast
     * which comes from client.
     */
    span.setOperationName(rpc);
}
```

Reply:
Why do you plan to propagate the op name in the wire? Entry span should focus on the server side logic only. The only thing will propagate in your RPC header, is the official ContextCarrier.

Like my other comments say, I have concern about you are hijacking the ContextCarrier. If we change it some day, even it is rarely to see, it is hard to remember the Finagle plugin has this operation.

Reply:
> Why do you plan to propagate the op name in the wire? 

For finagle version 17.12.0 and above,  we can not get op name in the server side directly.  If we add `finagle-{protocol}(eg: finagle-thrift)` into this plugin, then we could deserialize the bytes of the request paramenter and get op name, but this will introduce more complexity becase we have to be compatible with various protocols.

> Like my other comments say, I have concern about you are hijacking the ContextCarrier. If we change it some day, even it is rarely to see, it is hard to remember the Finagle plugin has this operation.

So far, I can't think of a better way to do this.  When we create span and contextCarrier, we don't know the remote address and op name,  but it is difficult to  propagate trace information across finagle's inner threads and then create span untill we know remote address and op name.

Reply:
Take a look at #4462